src/zarr/core/group.py

```diff
@@ -264,19 +559,51 @@ async def getitem(
         else:
             raise ValueError(f"unexpected zarr_format: {self.metadata.zarr_format}")
 
+    def _getitem_consolidated(self, store_path: StorePath, key: str) -> AsyncArray | AsyncGroup:
+        # getitem, in the special case where we have consolidated metadata.
+        # Note that this is a regular def (non async) function.
+        # This shouldn't do any additional I/O.
+
+        # the caller needs to verify this!
+        assert self.metadata.consolidated_metadata is not None
+
+        try:
+            metadata = self.metadata.consolidated_metadata.metadata[key]
+        except KeyError as e:
+            # The Group Metadata has consolidated metadata, but the key
+            # isn't present. We trust this to mean that the key isn't in
+            # the hierarchy, and *don't* fall back to checking the store.
+            msg = f"'{key}' not found in consolidated metadata."
+            raise KeyError(msg) from e
+
+        # update store_path to ensure that AsyncArray/Group.name is correct
+        store_path = StorePath(store=store_path.store, path=key)
+
+        if isinstance(metadata, GroupMetadata):
+            return AsyncGroup(metadata=metadata, store_path=store_path)
+        else:
+            return AsyncArray(metadata=metadata, store_path=store_path)
+
     async def delitem(self, key: str) -> None:
         store_path = self.store_path / key
         if self.metadata.zarr_format == 3:
             await (store_path / ZARR_JSON).delete()
+
         elif self.metadata.zarr_format == 2:
             await asyncio.gather(
                 (store_path / ZGROUP_JSON).delete(),  # TODO: missing_ok=False
                 (store_path / ZARRAY_JSON).delete(),  # TODO: missing_ok=False
                 (store_path / ZATTRS_JSON).delete(),  # TODO: missing_ok=True
             )
+
         else:
             raise ValueError(f"unexpected zarr_format: {self.metadata.zarr_format}")
 
+        if self.metadata.consolidated_metadata:
+            self.metadata.consolidated_metadata.metadata.pop(key, None)
+            # FIXME: This should probably rewrite all the consolidated metadata?
```

Comment:
How should we think about mutating operations (`setitem`, `delitem`) on `Group` (and Array) objects, and how they relate to the state in the backing `store`? And how does consolidated metadata factor in here?

My understanding is that mutating operations should be "immediately" reflected in the backing store, as part of the function call.

I *think* things are currently written such that the in-memory `Group` will always reflect the status of the *non*-consolidated metadata. If you do a `setitem` or `delitem`, it'll be set / deleted from both the in-memory Group and the backing store synchronously. Should we also try to write out the consolidated metadata on each operation?

Notably, we *shouldn't need to list each member* when *re*-consolidating the metadata. We should already know all the members thanks to the starting consolidated metadata and all the operations we've done on it (setting aside the case of multiple clients / Groups concurrently mutating the same store... that's a problem; we'll need a way to "force" reconsolidating).

I'm pretty conflicted here. I don't think there's an obviously correct answer here. The safest is *probably* to automatically re-consolidate though. Maybe the best option is a flag on the Group or Zarr config to control the behavior?

Reply:
Sorry @TomAugspurger this question didn't get answered sooner. Here's my perspective:

1. Mutating a consolidated store is inherently unsafe. In the case of multiple writers, there is not way ensure consistency without a lock/transaction layer in the store.
2. Readers should be able to opt in / out of trusting consolidated metadata. (some stores, like http, may require it)
3. Hierarchy and metadata mutating operations should probably invalidate the consolidated metadata.

Reply:
Thanks. I think I'm mostly aligned on that. Some clarification on 2 and 3:

1. If I open a Group, bypassing consolidated metadata with `use_consolidated=False`, I would expect mutating operations on that group to *not* touch the consolidated metadata (if any).
2. If I open a Group with consolidated metadata, should mutating operations *invalidate* the consolidated metadata (by deleting that entire key from the Group metadata) or *reconsolidate*?

Assuming we don't have concurrent writers mutating the same consolidated metadata (which we'll need to document), then invalidation and re-consolidation are both just a single atomic operation on the store (write the in-memory Group's metadata (including consolidated metadata) to the store).

Reply:
> If I open a Group with consolidated metadata, should mutating operations invalidate the consolidated metadata (by deleting that entire key from the Group metadata) or reconsolidate?

There is a 3rd option following a mutation -- do nothing! This is what we do in 2.x. I personally favor the `invalidate` option but we could also follow the behavior from 2.x and simply tell the user (through documentation or a warning) that they need to reconsolidate.

Reply:
> I personally favor the invalidate option

I'm coming around to that too. Initially, I favored "reconsolidate" since we can do that *without re-listing* so it would be as expensive as invalidating (which will just remove the `consolidated_metadata` key from the root zarr.json. But that is only safe to do when there's a single concurrent writer mutating the group.

The following would lead to inconsistent consolidated metadata:

```python
store = ...
await consolidate_metadata(store)

root = await open_group(store)
root2 = await open_group(store)

await asyncio.gather([root1.delitem("array-0"), root2.delitem("array-1")])
```

Whichever one ran last would "win", but the consolidated metadata would still contain the metadata for the other array.

So for now I think this is best solved through documentation. In the future, I hope that we can implement a better interface between the Store and Group that will allow for unit-of-work style updates to the metadata in-memory, and the ability to safely persist those in the store.